finality-aleph/src/data_io.rs

```diff
@@ -1,28 +1,62 @@
 use crate::{Error, Metrics};
 use aleph_bft::OrderedBatch;
-use futures::channel::mpsc;
+use futures::{
+    channel::{mpsc, oneshot},
+    stream::StreamExt,
+};
 use log::debug;
+use parking_lot::Mutex;
 use sp_consensus::SelectChain;
 use sp_runtime::traits::{Block, Header};
+use std::sync::Arc;
 
 #[derive(Clone)]
-pub(crate) struct DataIO<B: Block, SC: SelectChain<B>> {
-    pub(crate) select_chain: SC,
+pub(crate) struct DataIO<B: Block> {
+    pub(crate) best_chain: Arc<Mutex<B::Hash>>,
     pub(crate) ordered_batch_tx: mpsc::UnboundedSender<OrderedBatch<B::Hash>>,
+    pub(crate) refresh_best_chain_tx: mpsc::UnboundedSender<()>,
     pub(crate) metrics: Option<Metrics<B::Header>>,
 }
 
-impl<B: Block, SC: SelectChain<B>> aleph_bft::DataIO<B::Hash> for DataIO<B, SC> {
+pub(crate) async fn refresh_best_chain<B: Block, SC: SelectChain<B>>(
+    mut requests: mpsc::UnboundedReceiver<()>,
+    select_chain: SC,
+    best_chain: Arc<Mutex<B::Hash>>,
+    mut exit: oneshot::Receiver<()>,
+) {
+    loop {
+        tokio::select! {
+            _ = requests.next() => {
+                let new_best_chain = select_chain
+                    .best_chain()
+                    .await
+                    .expect("No best chain")
+                    .hash();
+                *best_chain.lock() = new_best_chain;
+            }
+            _ = &mut exit => {
+                debug!(target: "afa", "Task for refreshing best chain received exit signal. Terminating.");
+                return;
+            }
+        }
+    }
+}
+
+impl<B: Block> aleph_bft::DataIO<B::Hash> for DataIO<B> {
     type Error = Error;
 
     fn get_data(&self) -> B::Hash {
-        let header = self.select_chain.best_chain().expect("No best chain");
+        if self.refresh_best_chain_tx.unbounded_send(()).is_err() {
+            debug!(target: "afa", "Channel for new best chain requests is closed.");
```

Comment:
ymmv, but this used to result in a panic, now is a silent debug statement, maybe a `warning!` is in order?

Reply:
The diff does not reflect the corresponding part of the code. There is still a panic, but now it is in the line 33, in the refresh task.